Provide guest agent state based on lifecycle events

Ruclo opened the pull request:
This PR gets rid of the guest agent ping command execution in virtwrap manager and instead returns the state of the guest agent from the agent poller.
The agent poller stores the state of guest agent based on the guest agent lifecycle events received.
The guest ping can be used as a readiness probe for a vmi if configured by the user.

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as a draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### References
https://issues.redhat.com/browse/CNV-70177

### Special notes for your reviewer

Roses are red, commits are neat,
Thanks for the review, your feedback’s elite. :)

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least one e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

Ruclo committed:
refactor: Provide guest agent state based on lifecycle events

This commit gets rid of the guest agent ping command
execution in virtwrap manager and instead returns the
state of the guest agent from the agent poller.
The agent poller stores the state of guest agent based
on the guest agent lifecycle events received.
The goal of this change is to use the preferred libvirt api,
which is well tested

Signed-off-by: Michal Vavrinec <mvavrine@redhat.com>

akrejcir commented on pkg/virt-launcher/virtwrap/agent-poller/agent_poller.go:
```diff
@@ -45,6 +45,7 @@ const (
 	GetAgent          AgentCommand = "guest-info"
 	GetFSFreezeStatus AgentCommand = "guest-fsfreeze-status"
 
+	AgentConnected      = "agent-connected"
```
This constant is only used in this file, so it doesn't need to be exported.

Ruclo committed:
Remove redundant variable

When checking the qemu guest agent connection state, it's now stored in the AgentStore.
There is no need to keep the agentConnected boolean in the AgentPoller struct anymore.

Signed-off-by: Michal Vavrinec <mvavrine@redhat.com>

Ruclo commented on pkg/virt-launcher/virtwrap/manager.go:
```diff
@@ -154,7 +154,7 @@ type DomainManager interface {
 	GetGuestOSInfo() *api.GuestOSInfo
 	Exec(string, string, []string, int32) (string, error)
-	GuestPing(string) error
+	GuestPing() error
```
Should I rename it only at this level of abstraction (DomainManager) and keep it as GuestPing in other places (eg. the cmd server in virt-launcher)?

akrejcir replied:
Yes. The `cmd-server` implements the server side of a Protocol Buffer API, that is used for communication between `virt-handler` and `virt-launcher`. Renaming a method in the API will cause incompatibilities between different versions of these components. 
It could happen that a long running VM, that has older version of `virt-launcher` would not be able to communicate with a new `virt-handler` without restarting the VM.

Please only add a comment in the code that says that the method does not use guest ping anymore.

Maybe in a future PR, we can change the API in a backward compatible way.

0xFelix commented on pkg/virt-launcher/virtwrap/agent-poller/agent_poller.go:
```diff
@@ -220,6 +221,14 @@ func (s *AsyncAgentStore) GetLoad() *stats.DomainStatsLoad {
+	if !ok {
+		return false
+	}
+	return data.(bool)
```
You could check the assertion

Ruclo replied:
Even if the ok value from type assertion would be false, the underlying returned boolean would be false (zero value). is there any point in checking the assertion if I return false anyway?

0xFelix replied:
If it does return anyway then you can keep it.

0xFelix commented:
@akrejcir @Ruclo Does this relate to https://github.com/kubevirt/kubevirt/pull/16666?

0xFelix commented on pkg/virt-launcher/virtwrap/agent-poller/agent_poller.go:
```diff
@@ -384,13 +391,13 @@ func (p *AgentPoller) UpdateFromEvent(domainEvent *libvirt.DomainEventLifecycle,
 		if agentEvent.State == libvirt.CONNECT_DOMAIN_EVENT_AGENT_LIFECYCLE_STATE_DISCONNECTED {
 			log.Log.Infof("Stopping agent poller for %s due to agent disconnect", p.domainName)
-			p.agentConnected = false
+			p.agentStore.Store(agentConnected, false)
```
I don't know how the virtio ring is implemented, but apparently the hypervisor does not care about a crashing / disappearing guest agent.

A 'disconnect' event can only happen if the guest agent tears down gracefully. Do I understand that correctly? So we have to derive from failing queries / timeouts that the guest agent went away?
